zerver/models.py

```diff
@@ -2213,6 +2213,18 @@ def validate_attachment_request(user_profile: UserProfile, path_id: str) -> Opti
     # sent to a stream they are on where history is public to
     # subscribers.
 
+    relevant_stream_ids = messages.filter(
+        recipient__type=Recipient.STREAM).values_list("recipient__type_id", flat=True)
+
+    was_sent_to_web_public_stream = Stream.objects.filter(id__in=relevant_stream_ids,
+                                                          is_web_public=True).exists()
+
+    if (attachment.is_realm_public and attachment.realm == user_profile.realm and
+            was_sent_to_web_public_stream):
+        # Any user in the realm, including guest users, can access files sent
+        # to web-public streams.
+        return True
```

Comment:
This is a somewhat got code path, so we should be careful about adding more complexity to it.

I think we might be able to do this with fewer database queries by combining the logic with the next check.

It's also worth thinking about whether we want to add an `attachment.is_web_public` to sit alongside `is_realm_public`; that might be a better way to hook things up long-term, since when we add the main "is_web_public" feature of having content be public to the web, we might want to make it very cheap to do that check.

Reply:
I think adding `is_web_public` to the `attachment` is a better way to do it too. Could I go ahead and implement it? Or this would need more discussion?

Reply:
I think you should go for it.  Might be reasonable to just do that as an independent change on a new PR to make it easy to have CI run on it (I expect it'll involve some back-and-forth discussion).